src/assign.rs

```diff
@@ -313,26 +313,160 @@ mod json {
+                    ptr: "/1",
+                    value: json!("foo"),
+                    expected_data: json!({"1": "foo"}),
+                    expected_result: Ok(Some(json!(null))),
+                },
+                Test {
+                    data: json!([]),
+                    ptr: "/0",
+                    expected_data: json!(["foo"]),
+                    value: json!("foo"),
+                    expected_result: Ok(None),
+                },
+                Test {
+                    data: json!([]),
+                    ptr: "/0",
+                    expected_data: json!(["foo"]),
+                    value: json!("foo"),
+                    expected_result: Ok(None),
+                },
+                Test {
+                    data: json!([]),
+                    ptr: "/-",
+                    value: json!("foo"),
+                    expected_result: Ok(None),
+                    expected_data: json!(["foo"]),
+                },
+                Test {
+                    data: json!([]),
+                    ptr: "/1",
+                    value: json!("foo"),
+                    expected_result: Err(AssignError::OutOfBounds {
+                        offset: 0,
+                        source: OutOfBoundsError{
+                            index: 1,
+                            length: 0,
+                        },
+                    }),
+                    expected_data: json!([]),
+                },
+                Test {
+                    data: json!([]),
+                    ptr: "/a",
+                    value: json!("foo"),
+                    expected_result: Err(AssignError::FailedToParseIndex {
+                        offset: 0,
+                        source: json::ParseIndexError { source: usize::from_str("foo").unwrap_err() },
+                    }),
+                    expected_data: json!([]),
+                },
+            ];
 
-            let tests = [(json!({}), "/foo", json!("bar"))];
-            for (mut data, ptr, value) in tests {
+            for (i,Test {
+                mut data,
+                ptr,
+                value,
+                expected_data,
+                expected_result: expected_replaced,
+            }) in tests.into_iter().enumerate()
+            {
                 let ptr = PointerBuf::from_str(ptr).unwrap();
-                let replaced = ptr.assign(&mut data, value);
+                let replaced = ptr.assign(&mut data, value.clone());
+                assert_eq!(
+                    &expected_data, &data,
+                    "data not as expected.\n\nexpected:\n\n{expected_data}\n\nactual:\n{data}\n"
+                );
+                assert_eq!(
+                    &expected_replaced,
+                    &replaced,
+                    "\nreplaced not as expected for test index {i}:\n\npointer:{ptr}\nvalue:{value}\n\ndata:\n{data}\n\nexpected data:\n{expected_data}\n\nexpected:\n{}\n\nactual:\n{}\n\n",
```

Comment:
I still think `insta` is worth using here. Keeping this up-to-date manually is a pain, and it's hard to read.

Reply:
I wasn't sure how obvious it was to tell the expectations of `insta` vs this which you can see the input vs expected output.

Reply:
The `insta` paradigm is that you never explicitly write the expected output, it will run your test once and record the output in a file, which becomes the ground truth. From that point on, every run of the test will use the previously accepted output, so you get the same effect as using `assert_eq!`, but without cluttering the code with these strings. Then if you ever change the expected output, you can just regenerate the snapshot.

Reply:
Right, I did watched the video and read through some of the docs.

I can understand that table-testing is funky to read at first. I absolutely did not care for it when I first saw it. It took exposure and experimentation to finally come around to it.

Setting insta aside, the advantages to it are being able to see the cases covered and expected output at a glance. Ideally helping facilitate tests as docs. They are a pain to setup but ironically allow for pretty fast coverage and edge-case correction. It just takes some getting used to. There is a bit of boilerplate though, so it does clutter things up I guess.

Are they not common in rust? Is insta favored enough that people know to run or review the cases? I'll give it a shot.

Also, sorry about the strings. They were there largely to help me debug as I went through the cases.

Reply:
I'm not sure we're on the same page, table-testing and insta aren't either/or at all! But it's my fault, I commented in the wrong place, the hard-to-read string here is not actually the problem, since it's just providing a nicer error message if the test fails, rather than the error message from an `Err` you're expecting. Arguably the solution to make this more readable then is using something like [`indoc`](https://docs.rs/indoc/latest/indoc/) instead. Sorry for the mixup. 

`insta` may still be useful here so that you don't have to explicitly craft `expected_result` and `expected_data`. To be super clear, you'd still have table-testing, that doesn't change at all. It's just how you define the expected output that changes.

I've proposed this because I've found it helps with long term maintenance, requiring less manual intervention if the internal structure of errors change or as new test cases are added, but given how you've defined errors transparently, perhaps it's not worth it. It shines the most when errors are opaque, or you expect complex data that's tricky to craft manually (e.g., a deeply nested JSON). So if you prefer to keep things straightforward, that's totally fine.

Reply:
Ahh, okay. I'm game for using `insta`. It's a super cool tool, from the looks of it.

Thank you for taking the time to explain!